File: 2022-05/python/21_Merge_Two_Sorted_Lists.py
class Solution:
    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        return self.recursively(l1, l2)
        return self.iteratively(l1, l2)

    def recursively(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if not l1 or not l2:
            return l1 or l2
        if l1.val < l2.val:
            l1.next = self.recursively(l1.next, l2)
            return l1
        else:
            l2.next = self.recursively(l1, l2.next)
            return l2

    def iteratively(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        cur = start = ListNode()
        while l1 and l2:
            if l1.val < l2.val:
                cur.next = l1
                l1 = l1.next
            else:
                cur.next = l2
                l2 = l2.next
            cur = cur.next
        cur.next = l1 or l2
        return start.next

File: 2022-05/javascript/21_Merge_Two_Sorted_Lists.js
function mergeTwoLists(l1, l2) {
    // return recursively(l1, l2)
    return iteratively(l1, l2)
}

function iteratively(l1, l2) {
    const start = new ListNode()
    let cur = start
    while (l1 && l2) {
        if (l1.val < l2.val) {
            cur.next = l1
            l1 = l1.next
        } else {
            cur.next = l2
            l2 = l2.next
        }
        cur = cur.next
    }
    cur.next = l1 ? l1 : l2
    return start.next
}

function recursively(l1, l2) {
    if (!l1 || !l2) {
        return l1 | l2
    }
    if (l1.val < l2.val) {
        l1.next = recursively(l1.next, l2)
        return l1
    } else {
        l2.next = recursively(l1, l2.next)
        return l2
    }
}